zerver/views/registration.py

```diff
@@ -881,6 +886,75 @@ def redirect_to_email_login_url(email: str) -> HttpResponseRedirect:
     return HttpResponseRedirect(redirect_url)
 
 
+@typed_endpoint
+def realm_import_status(
+    request: HttpRequest,
+    maybe_user_profile: UserProfile | AnonymousUser,
+    *,
+    confirmation_key: str,
+) -> HttpResponse:
+    try:
+        # We don't use get_object_from_key because we're fine if it's been marked used
+        confirmation = Confirmation.objects.get(
+            confirmation_key=confirmation_key, type__in=[Confirmation.REALM_CREATION]
+        )
+    except Confirmation.DoesNotExist:
+        raise JsonableError(_("Unauthenticated"))
+
+    preregistration_realm = confirmation.content_object
+    assert isinstance(preregistration_realm, PreregistrationRealm)
+    try:
+        realm = Realm.objects.get(string_id=preregistration_realm.string_id)
+    except Realm.DoesNotExist:
+        # XXX Seplunk the filesystem for the converted import to snoop
+        # on which files have been written?  Only one dir should match
+        # the tempdir glob
+        return json_success(request, {"status": _("Converting Slack data…")})
+
+    if realm.deactivated:
+        # These "if" cases are in the inverse order than they're done
+        # in the import process, so we get the latest step that it's
+        # on.
+        if Message.objects.filter(realm_id=realm.id).exists():
+            return json_success(request, {"status": _("Importing messages…")})
+        try:
+            next(all_message_attachments(prefix=f"{realm.id}/"))
+            return json_success(request, {"status": _("Importing attachment data…")})
+        except StopIteration:
+            pass
+        return json_success(request, {"status": _("Importing converted Slack data…")})
+
+    if (
+        not preregistration_realm.data_import_metadata["no_user_matching_email"]
+        and preregistration_realm.created_realm is None
+    ):
+        return json_success(request, {"status": _("Finalizing import…")})
+
+    # We have a non-deactivated realm and it's linked to the prereg key
+    result = {"status": _("Done!")}
+    if not preregistration_realm.data_import_metadata["no_user_matching_email"]:
+        original_user = UserProfile.objects.get(
+            delivery_email__iexact=preregistration_realm.email, realm=realm
+        )
```

Comment:
This can be `get_user_by_delivery_email`, right?

Reply:
Yeah. changed